include/can/core/arbitration_id.hpp

```diff
@@ -5,33 +5,79 @@
 namespace can_arbitration_id {
 
 /**
- * The components of a 29-bit arbitration id as a bitfield.
+ * A class that represents our 29-bit arbitration id.
  */
-struct ArbitrationIdParts {
-    unsigned function_code : 7;
-    unsigned node_id : 8;
-    unsigned message_id : 14;
-    unsigned int padding : 3;
-};
+class ArbitrationId {
+  public:
+    ArbitrationId() = default;
+    ArbitrationId(uint32_t arbitration_id) : id{arbitration_id} {}
 
-/**
- * A union of arbitration id in parts or as an integer.
- */
-union ArbitrationId {
-    ArbitrationIdParts parts;
-    uint32_t id;
-};
+    /**
+     * Implicit conversion to uint32_t
+     * @return the id
+     */
+    operator uint32_t() const { return get_id(); }
 
-/**
- * Message id mask.
- */
-constexpr auto message_id_mask = ArbitrationId{.parts{
-    .function_code = 0, .node_id = 0, .message_id = 0x3FFF, .padding = 0}};
+    /**
+     * Get the encoded arbitration id as an integer
+     * @return the id
+     */
+    [[nodiscard]] auto get_id() const -> uint32_t { return id; }
 
-/**
- * Node id mask.
- */
-constexpr auto node_id_mask = ArbitrationId{
-    .parts{.function_code = 0, .node_id = 0xFF, .message_id = 0, .padding = 0}};
+    [[nodiscard]] auto function_code() const -> uint8_t {
+        return get(function_code_bit_mask, function_code_shift);
+    }
+    [[nodiscard]] auto node_id() const -> uint8_t {
+        return get(node_id_bit_mask, node_id_shift);
+    }
+    [[nodiscard]] auto message_id() const -> uint16_t {
+        return get(message_id_bit_mask, message_id_shift);
+    }
+
+    void function_code(uint8_t v) {
```

Comment:
These IMO should take enum members rather than `uint8`s. It reinforces that these are _not_ general ints but specific values, and it means we don't have to clutter up call sites with `static_cast`s.